src/rdf4cpp/rdf/util/SharedPtr.hpp

```diff
@@ -0,0 +1,524 @@
+#ifndef RDF4CPP_RDF_UTIL_SHAREDPTR_HPP
+#define RDF4CPP_RDF_UTIL_SHAREDPTR_HPP
+
+#include <atomic>
+#include <cassert>
+#include <utility>
+#include <optional>
+
+#include <rdf4cpp/rdf/storage/util/robin-hood-hashing/robin_hood_hash.hpp>
+
+namespace rdf4cpp::rdf::util {
+
+namespace shared_ptr_detail {
+
+/**
+ * The control block of SharedPtr/WeakPtr
+ */
+struct ControlBlock {
+    /**
+     * Number of SharedPtrs (pointing to this control block) currently alive
+     * Values greater than 0 keep the data (not stored here) alive.
+     */
+    std::atomic<size_t> strong;
+
+    /**
+     * Number of WeakPtrs (pointing to this control block) currently alive.
+     * Values greater than 0 keep this control block alive.
+     */
+    std::atomic<size_t> weak;
+};
+
+}  //namespace shared_ptr_detail
+
+template<typename T>
+struct WeakPtr;
+
+template<typename T>
+struct SharedPtr;
+
+/**
+ * Equivalent to std::enable_shared_from_this<T>
+ */
+template<typename T>
+struct EnableSharedFromThis {
+private:
+    template<typename U>
+    friend struct SharedPtr;
+
+    WeakPtr<T> mutable weak;
+
+    void set_owner(WeakPtr<T> &&owner) const noexcept {
+        this->weak = std::move(owner);
+    }
+public:
+    SharedPtr<T> shared_from_this() const noexcept {
+        return this->weak.upgrade();
+    }
+
+    WeakPtr<T> weak_from_this() const noexcept {
+        return this->weak;
+    }
+};
+
+/**
+ * Shared pointer implementation, that exposes ptr equality and hash on WeakPtrs.
+ * This implementation is inspired by Rust's std::sync::Arc and boost's boost::shared_ptr.
+ * @tparam T type of object to manage
+ */
+template<typename T>
+struct SharedPtr {
```

Comment:
I honestly don't really want to have a SharedPtr implementation in here but for _some_ reason `std::weak_ptr` and `boost::weak_ptr` both do not expose pointer equality even though that is a perfectly valid operation
on the control block (because weak pointer keep that alive and there is a 1-1 relationship between managed objects and control blocks).

This functionality is needed for fast (and user friendly) comparisions of BlankNodes in merge and union semantics.

This is needed here: https://github.com/rdf4cpp/rdf4cpp/blob/1a7b18ddd4f9c132534d341de7abc300828200cb/src/rdf4cpp/rdf/BlankNode.cpp#L37-L76

Reply:
Makes sense to me. As discussed, a through review might need some second opinion from https://codereview.stackexchange.com/ after it was ported to offset pointers. :)